pyop2/caching.py

```diff
@@ -240,14 +241,109 @@ def cache_key(self):
 .. note::
     If you intend to use this decorator to cache things that are collective
     across a communicator then you must include the communicator as part of
-    the cache key. Since communicators are themselves not hashable you should
-    use :func:`pyop2.mpi.hash_comm`.
+    the cache key.
 
     You should also make sure to use unbounded caches as otherwise some ranks
     may evict results leading to deadlocks.
 """
 
 
+def default_parallel_hashkey(*args, **kwargs):
+    comm = kwargs.get('comm')
+    return comm, cachetools.keys.hashkey(*args, **kwargs)
+
+
+def parallel_memory_only_cache(key=default_parallel_hashkey):
+    """Memory only cache decorator.
+
+    Decorator for wrapping a function to be called over a communiucator in a
+    cache that stores broadcastable values in memory. If the value is found in
+    the cache of rank 0 it is broadcast to all other ranks.
+
+    :arg key: Callable returning the cache key for the function inputs. This
+        function must return a 2-tuple where the first entry is the
+        communicator to be collective over and the second is the key. This is
+        required to ensure that deadlocks do not occur when using different
+        subcommunicators.
+    """
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            """ Extract the key and then try the memory cache before falling back
+            on calling the function and populating the cache.
+            """
+            comm, mem_key = key(*args, **kwargs)
```

Comment:
Maybe it would be "cleaner" to have 2 callables, one returning the comm and the other the hash key. The 2-tuple approach was needed for the `collective` kwarg to work I think.